crates/sui-framework/packages/sui-framework/sources/registries/coin_metadata_registry.move

```diff
@@ -0,0 +1,273 @@
+    @0x10.to_id()
+}
+
+// === Coin Metadata ===
+public struct CoinMetadataKey<phantom T>() has copy, drop, store;
+
+public struct MetadataCap<phantom T> has key, store { id: UID }
+
+public struct Metadata<phantom T> has key, store {
+    id: UID,
+    decimals: u8,
+    name: String,
+    symbol: String,
+    description: String,
+    icon_url: String,
+    supply: Option<Supply<T>>,
+    treasury_cap_id: Option<ID>,
+    metadata_cap_id: Option<ID>,
+    deny_cap_id: Option<ID>,
+}
+
+// hot potato pattern to enforce registration after "new_currency" metadata creation
+public struct InitMetadata<phantom T> {
+    metadata: Metadata<T>,
+}
+
+// call this after create_currency to register the metadata
+public fun transfer<T>(init: InitMetadata<T>) {
+    let InitMetadata { metadata } = init;
+
+    transfer::public_transfer(
+        metadata,
+        coin_metadata_registry_id().to_address(),
+    );
+}
+
+/// This is for TTO registration flow.
+public fun migrate_receiving<T>(
+    registry: &mut CoinMetadataRegistry,
+    metadata: Receiving<Metadata<T>>,
+) {
+    let received_metadata = transfer::public_receive(&mut registry.id, metadata);
+    registry.attach_metadata(received_metadata);
+}
+
+public fun create_cap_for_supply<T>(
+    registry: &CoinMetadataRegistry,
+    _supply: &mut Supply<T>,
+    publisher: &Publisher,
+    ctx: &mut TxContext,
+): MetadataCap<T> {
+    // only coin metadata objects can be registered
+    assert!(publisher.from_package<T>(), ENotSystemAddress);
+    let metadata = registry.metadata<T>();
+    assert!(!metadata.cap_claimed(), EAlreadyClaimed);
+    MetadataCap { id: object::new(ctx) }
+}
+
+/// === Setters ===
+
+public fun set_name<T>(_: &MetadataCap<T>, registry: &mut CoinMetadataRegistry, name: String) {
+    registry.metadata_mut<T>().name = name;
+}
+
+public fun set_symbol<T>(_: &MetadataCap<T>, registry: &mut CoinMetadataRegistry, symbol: String) {
+    registry.metadata_mut<T>().symbol = symbol;
+}
+
+public fun set_description<T>(
+    _: &MetadataCap<T>,
+    registry: &mut CoinMetadataRegistry,
+    description: String,
+) {
+    registry.metadata_mut<T>().description = description;
+}
+
+public fun set_icon_url<T>(
+    _: &MetadataCap<T>,
+    registry: &mut CoinMetadataRegistry,
+    icon_url: String,
+) {
+    registry.metadata_mut<T>().icon_url = icon_url;
+}
+
+public(package) fun freeze_supply<T>(registry: &mut CoinMetadataRegistry, supply: Supply<T>) {
+    registry.metadata_mut<T>().supply.fill(supply);
+}
+
+public(package) fun set_decimals<T>(registry: &mut CoinMetadataRegistry, decimals: u8) {
+    registry.metadata_mut<T>().decimals = decimals;
+}
+
+public(package) fun set_deny_cap<T>(metadata: &mut Metadata<T>, deny_cap_id: Option<ID>) {
+    metadata.deny_cap_id = deny_cap_id;
+}
+
+/// === Getters ===
+
+public fun metadata<T>(registry: &CoinMetadataRegistry): &Metadata<T> {
+    assert!(registry.exists<T>(), EMetadataNotFound);
+    dynamic_field::borrow(
```

Comment:
You probably (IMO) want to use a `dynamic_object_field` instead of `dynamic_field` here and for holding the `Metadata`.